[2단계 - 점진적인 리팩터링] 노랑 미션 제출합니다

yesjuhee opened the pull request:
## 기능 요구사항 구현 & 테스트

Controller 인터페이스 기반의 레거시 MVC와, 1단계에서 추가된 @Controller 어노테이션 기반의 MVC가 함께 작동하는 구조를 구현했습니다.

해당 기능을 실제로 테스트 해보기 위해 `Step2Controller` 컨트롤러를 추가했습니다. 서버 실행하고 `localhost:8080/form` 으로 요청을 보내면 해당 컨트롤러 로직을 테스트할 수 있습니다!

아래 3개 파일은 테스트용으로 추가한거라 리뷰에서 제외하고 보시면 됩니다
- `app/src/main/webapp/result.jsp`
- `app/src/main/webapp/form.jsp`
- `app/src/main/java/com/techcourse/controller/Step2Controller.java`

## 추가된 클래스

기존 기능을 유지하면서, 신규 기능을 추가하기 쉬운 구조를 만들기 위해 LMS에서 제시된 구조를 최대한 따라가는 방식으로 진행했습니다.

### ControllerScanner
기존 `AnnotationHandlerMapping`에서 하던 일부 작업을 가져갔습니다. 크게 필요성을 느끼진 못했는데 일단 LMS 구조를 최대한 따라가보는 방향으로 진행했습니다.

### HandlerMapping, HandlerMappingRegistry

`AnnotationHandlerMapping`, `ManualHandlerMapping` 처럼 Http Request에 해당하는 적절한 헨들러를 반환하는 클래스를 `HandlerMapping` 인터페이스로 추상화하여 관리합니다. 

MVC에서 사용되는 모든 HandlerMapping을 `HandlerMappingRegistry`의 생성자에서 초기화하여 관리하고 있습니다.

### HandlerAdapter, HandlerAdapterRegistry

핸들러의 종류에 따라서 작동 방식이 다르기 때문에 이를 추상화 하기 위해 `HandlerAdapter` 인터페이스를 도입했습니다. 해당 인터페이스는 `org.springframework.web.servlet.HandlerAdapter`를 참고했습니다.

MVC에서 사용되는 모든 HandlerMapping을 `HandlerAdapterRegistry`의 생성자에서 초기화하여 관리하고 있습니다.

## 임의로 변경한 부분

주어진 클래스 다이어그램을 최대한 따라가긴 했는데, 몇가지 잘 이해가 안되는 부분은 임의로 변경했습니다.

- `DispatcherServlet::init` 메서드 삭제
- `HandlerAdapterRegistry::addHandlerAdapter`, `HandlerMappingRegistry::addHandlerAdapter` 메서드를 private 으로 변경해서 사용

처음에는 기존 구조에 최대한 맞춰서 `DispatcherServlet::init` 메서드에서 `addXxx` 를 써서 `xxxRegistry` 클래스에 객체들을 추가하는 방식으로 구현했습니다. 그런데 DispatcherServlet에 너무 많은 책임이 있는 것 같아서 각 Registry 클래스가 등록 목록들을 관리하도록 변경했습니다. 그래서 `DispatcherServlet::init` 메서드가 필요 없어졌고, `addXxx` 메서드도 내부에서만 사용하기 때문에 private 으로 변경했습니다.

생성에 대한 책임은 최대한 생성자에 있는걸 선호하는 편인데, 이번 뼈대 코드는 거의 빈 생성자를 사용해서 좀 의아하게 생각하고 있어요ㅎㅎ

- `HandlerMappingRegistry::getHandler`의 반환 타입을 `Opitonal<Object>` 대신 `Object` 사용

`xxRegistry` 클래스의 `getXxx` 메서드는 둘 다 없으면 내부에서 예외를 터트리도록 수정했습니다. Optional로 구현하면, 존재하지 않을 경우의 예외 처리를 `DispatcherServlet`에서 해줘야 하는데 위와 마찬가지로 이것도 `DispatcherServelt`의 관심사가 아니라고 판단했습니다.


혹시 이와 관련해서 추가적인 의견 있으면 듣고 싶어요 😄

yesjuhee committed:
refactor: AnnotationHandlerMapping 코드 리팩터링

- ControllerScanner 추가
- 개행 맞춤
- 필드 이름 변경

yesjuhee committed:
feat: 어노테이션 기반 컨트롤러 기능 추가

- `DispatcherServlet`에서 핸들러 매핑 처리 로직 수정
- 새로운 인터페이스 `HandlerMapping` 정의 및 구현

yesjuhee committed:
test: 어노테이션 컨트롤러 테스트 추가

- 어노테이션 기반 컨트롤러를 통해 폼 페이지(`/form`)와 결과 페이지(`/result`) 처리 기능 구현
- `Step2Controller` 추가하여 GET 및 POST 요청 대응
- `form.jsp` 및 `result.jsp` 뷰 파일 생성하여 사용자 입력 및 결과 표시

yesjuhee committed:
refactor: 핸들러 어댑터 패턴 추가 및 구조 개선

- `HandlerAdapter` 인터페이스 및 구현체 추가 (`ManualHandlerAdapter`, `AnnotationHandlerAdapter`)
- 기존 핸들러 매핑 및 레지스트리를 리팩터링하여 어댑터 패턴 적용
- `DispatcherServlet`에서 핸들러와 어댑터를 활용해 요청 처리 로직 개선
- 불필요한 메서드 제거 및 코드 간소화

yesjuhee committed:
refactor: 패키지 구조 변경 및 관련 코드 정리

- 필요 없는 테스트 메서드 코드 제거

yesjuhee commented:
@coderabbitai review
instructions 기반으로 리뷰~

yesjuhee commented on mvc/src/main/java/com/interface21/webmvc/servlet/mvc/tobe/handler/mapping/AnnotationHandlerMapping.java:
```diff
@@ -27,35 +26,23 @@ public AnnotationHandlerMapping(final Object... basePackage) {
 
-    public Object getHandler(final HttpServletRequest request) {
+    @Override
+    public HandlerExecution getHandler(final HttpServletRequest request) {
         final var requestURI = request.getRequestURI();
         final var method = request.getMethod();
         final var handlerKey = new HandlerKey(requestURI, RequestMethod.valueOf(method));
         return handlerExecutions.get(handlerKey);
     }
```
제시된 jsp 파일을 기준으로, 지금 처리할 수 있는 에러는 401, 404, 500만 있다고 가정하고 있어서 이 리뷰는 반영이 어려울 것 같음
@coderabbitai

yesjuhee committed:
fix: 리플렉션 호출 시 접근 제어 설정 수정

- `method.setAccessible(true)` 추가하여 접근 제어 문제 해결
- 리플렉션 호출 과정에서 발생할 수 있는 잠재적 예외 방지 및 안정성 강화

yesjuhee commented on mvc/src/main/java/com/interface21/webmvc/servlet/mvc/tobe/handler/mapping/HandlerExecution.java:
```diff
@@ -7,15 +7,15 @@
         this.method = method;
     }
 
     public ModelAndView handle(final HttpServletRequest request, final HttpServletResponse response) throws Exception {
-        return (ModelAndView) method.invoke(handler, request, response);
+        return (ModelAndView) method.invoke(declaredObject, request, response);
     }
```
ea12e6da

@coderabbitai 너 아주 핵심을 찔렀어

Byesol commented on app/src/main/java/com/techcourse/controller/Step2Controller.java:
```diff
@@ -0,0 +1,33 @@
+import jakarta.servlet.http.HttpServletResponse;
+
+@Controller
+public class Step2Controller {
```
실행 잘 되네요! 직접 controller와 jsp를 만들어 테스트해보신 것 좋네요!

Byesol commented on app/src/main/java/com/techcourse/HandlerMappingRegistry.java:
```diff
@@ -0,0 +1,40 @@
+                .filter(Objects::nonNull)
+                .findFirst()
+                .orElseThrow(
+                        () -> new ServletException("No handler found for request URI : " + request.getRequestURI())
```
-  HandlerMappingRegistry::getHandler의 반환 타입을 Opitonal<Object> 대신 Object 사용, xxRegistry 클래스의 getXxx 메서드는 둘 다 없으면 내부에서 예외를 터트리도록 수정했습니다. Optional로 구현하면, 존재하지 않을 경우의 예외 처리를 DispatcherServlet에서 해줘야 하는데   위와 마찬가지로 이것도 DispatcherServelt의 관심사가 아니라고 판단했습니다.

---

라고 하셨는데 핸들러 없음(404)과 어댑터 없음(500) 은 다른 문제 아닐까요?

getHandlerAdapter() 실패 (어댑터 없음) 가 발생한  시점은 이미 유효한 핸들러를 찾은 후입니다. 그럼에도 어댑터가 없다는 것은 서버 설정 오류가 맞기 때문에 오류를 던져 요청을 즉시 중단시키고 노랑이 한대로 500 응답을 던지는 게 맞을 것 같습니다.

그러나 getHandler() 실패 (핸들러 없음) 가 발생한 경우  클라이언트(사용자)가 존재하지 않는 URL을 요청한 경우입니다. 이는 서버의 버그가 아닌 정상적인 실패이지 않을까요? 현재 코드상으로는 바로 ServletException를 던지는데 이 경우 상위의 서블릿 컨테이너가 내부 오류로 판단해 곧바로 500 응답을 생성해서 클라이언트에게 보내주지 않을까요?

이 때, 클라이언트에게 404 응답을 보내려면 어떻게 해야 할까요?

yesjuhee replied:
너무 좋은 리뷰네요! 예외 처리까진 생각하지 못했었습니다.
`NoHandlerFoundException`을 던지면 `DispatcherServlet`에서 404로 에러 응답을 보내주도록 처리했습니다!

12caa6da

Byesol replied:
404 에러 응답을 처리할 수 있게 바꾼 점 좋네요!

Byesol commented on mvc/src/main/java/com/interface21/webmvc/servlet/mvc/tobe/handler/mapping/ControllerScanner.java:
```diff
@@ -0,0 +1,49 @@
+
+    private Object instantiateController(Class<?> controllerClass) {
+        try {
+            return controllerClass.getDeclaredConstructor().newInstance();
```
getDeclaredConstructor를 통해 private 생성자를 가져와도 setAccesible(true) 처리가 없기 때문에 private 생성자는 생성 못하지 않을까요?

yesjuhee replied:
68062975

앗 이 부분을 자꾸 놓치네요! 꼼꼼한 리뷰 감사합니다 👀

Byesol commented on app/src/main/java/com/techcourse/HandlerMappingRegistry.java:
```diff
@@ -0,0 +1,40 @@
+        manualHandlerMapping.initialize();
+        adHandlerMapping(manualHandlerMapping);
+
+        final var annotationHandlerMapping = new AnnotationHandlerMapping();
```
AnnotationHandlerMapping 의 생성자는 basePackage를 설정할 수도 있는데 여기서 basePackage를 설정하지 않은 건 어떤 이유 때문인가요?

yesjuhee replied:
예리한 질문이네요ㅎㅎ

### basePackage 를 설정하지 않은 이유

프레임워크 영역에서 어플리케이션의 상세 구현 여부를 알게 되는 것이라고 판단해서 설정하지 않았습니다.
지금은 구조의 한계상 `HandlerMappingRegistry`가 app 패키지에 속하지만, 결국 `mvc` 패키지에 속하도록 하는 것이 최종 구조라고 생각했어요!

즉 `HandlerMappingRegistry`는 mvc 프레임워크의 영역인데, 애플리케이션 영역인 `com.techcourse`를 알고 있고, 설정하는 것이 어색하다고 판단해서 설정하지 않았습니다.

### 실제 Springboot mvc 프레임워크는?

실제 springboot mvc 프레임워크에서는 `@ComponentScan` 어노테이션을 사용해서 reflection을 사용할 basePackage를 지정한다고 합니다. `@SpringBootApplication` 어노테이션에 `@ComponentScan`이 포함되어 있어서, 어플리케이션 클래스가 속한 패키지와 그 하위 패키지를 스캔 대상으로 설정하는 방법을 씁니다.

### 중간 타협점

고민을 하다가 미션을 위한 특수한 상황이라고 가정하고 패키지 명을 추가했습니다ㅎㅎ
df38e0bc

컴포넌트 스캔 기능 까지 구현한다면 best 겠지만 시간상 무리가 있어서 중간 타협 지점으로 패키지 명을 설정했습니다 👍

Byesol replied:
프레임워크 영역에서 어플리케이션의 상세 구현 여부를 알지 못하게 하려 한 노랑의 의도가 너무 좋네요.

단순히 패키지명을 바로 쓰는 것과 , 위와 같은 생각을 했지만, 미션을 위한 특수 상황임을 인정하고 패키지명을 쓰는 건 천지차이라고 생각해요. 

무엇이 더 나은 설계인지 명확히 인지하고 그걸 설명해주신 덕분에 많이 배우고 갑니다!

Byesol reviewed (changes requested):
안녕하세요! 노랑! 2단계도 테스트 컨트롤러까지 포함해 꼼꼼하게 코드를 작성해주셨네요! 

클래스들이 잘 분리되어 있고, 친절한 설명도 써셔서 이해가 잘 됐습니다.  ദ്ദി*ˊᗜˋ*)

몇 가지 커멘트 및 남겨주신 것들에 대한 제 주관적인 생각들은 커멘트로 남겨뒀습니다!!

Byesol commented on app/src/main/java/com/techcourse/HandlerMappingRegistry.java:
```diff
@@ -0,0 +1,40 @@
+import java.util.List;
+import java.util.Objects;
+
+public class HandlerMappingRegistry {
```
## Registry의 책임과 DispatcherServlet::init

> DispatcherServlet에 너무 많은 책임이 있는 것 같아서 각 Registry 클래스가 등록 목록들을 관리하도록 변경했습니다.

라고 하셨는데 DispatcherServlet의 책임이 과도하다고 느끼고 변경한 건 객체지향적으로 고민 끝에 나온 좋은 설계 같아요.

 그런데 현재 구조는 Registry가 ManualHandlerMapping이나 AnnotationHandlerAdapter 같은 구체적인 클래스를 직접 생성(new)하고 등록하는 책임까지 갖게 되었는데요

만약 기존 레거시 컨트롤러를 제거하거나 새로운 컨트롤러를 추가하려면 HandlerMappingRegistry와 HandlerAdapterRegistry 양쪽의 내부 코드를 모두 수정해야 하는데
이는 곧 새로운 컨트롤러를 도입할 때 Registry 내부 구현을 알아야 한다는 뜻 아닐까요?

그리고 DispatcherServlet 에 책임이 과도하게 몰린다면 DispatcherInitializer에 생성하고 등록하는 책임을 맡길 수 있지 않을까요?

yesjuhee replied:
### 내부 구현을 알아야 함

> 이는 곧 새로운 컨트롤러를 도입할 때 Registry 내부 구현을 알아야 한다는 뜻 아닐까요?

맞습니다. 지금 구조도 역시 문제가 있는 구조입니다 😂 
미션 코드의 구조 상 새로운 종류의 Handler가 등장하면 내부 코드를 건드릴 수 밖에 없습니다!
- HandlerMapping과 HandlerAdapter 생성 책임을 `DispatcherServlet`에 둘 경우 `DispatcherServlet`을 수정해야 합니다.
- HandlerMapping과 HandlerAdapter 생성 책임을 각각의 `Registry`에 둘 경우 각 `Registry`를 수정해야 합니다.

어느쪽에 등록되든 추가되어야 하는 코드의 양은 동일할 것 같습니다. `DispatcherServlet`에서 다 처리한다면 한 곳에서 수정할 수 있어서 장점이 있고, 각각의 `Registry`에서 처리한다면 등록을 할 때 각 책임에 맞는 부분만 신경쓰면 된다는 점에서 장점이 있습니다.

저는 후자의 장점이 더 크다고 판단해서 생성 책임을 분리했습니다!

### DispatcherInitializer

> DispatcherInitializer에 생성하고 등록하는 책임을 맡길 수 있지 않을까요?

해당 객체는 DispatcherServlet 을 초기화 하는 클래스이기 때문에 HandlerMapping, HandlerAdapter 에 대한 책임이 필요 없다고 판단했습니다.

---

### 그럼 Springboot MVC는?

그럼 실제 스프링부터 프레임워크에서 핸들러 매핑은 어떻게 불러올까요?! 이 부분이 궁금해져서 찾아봤습니다.

`DispatcherServlet`의 `initHandlerMappings`라는 메서드를 보면 다음과 같이 `BeanFactoryUtils`에서 가져오고 있습니다. 즉 HandlerMapping이 Bean으로 등록되어 있습니다.

```java
private void initHandlerMappings(ApplicationContext context) {
	this.handlerMappings = null;

	if (this.detectAllHandlerMappings) {
		// Find all HandlerMappings in the ApplicationContext, including ancestor contexts.
                // *** HandlerMapping 클래스 타입으로 빈을 찾아온다 ***
		Map<String, HandlerMapping> matchingBeans =
				BeanFactoryUtils.beansOfTypeIncludingAncestors(context, HandlerMapping.class, true, false); 
		if (!matchingBeans.isEmpty()) {
			this.handlerMappings = new ArrayList<>(matchingBeans.values());
			// We keep HandlerMappings in sorted order.
			AnnotationAwareOrderComparator.sort(this.handlerMappings);
		}
	}
}
```

그럼 빈은 어디서 등록이 되는걸까요? 또 찾아보니까 `WebMvcConfigurationSupport`에서 빈으로 등록해주고 있습니다. 그런데 신기한 부분은 **모든 HandlerMapping, HandlerAdapter를 직접 등록**하고 있더라구요!!

```java
public class WebMvcConfigurationSupport implements ApplicationContextAware, ServletContextAware {
...
@Bean
public RequestMappingHandlerMapping requestMappingHandlerMapping(...) {
    // @RequestMapping을 처리하는 HandlerMapping을 생성하고 빈으로 등록
    RequestMappingHandlerMapping mapping = createRequestMappingHandlerMapping();
    // ... 여러 설정 적용 ...
    return mapping;
}

@Bean
public HandlerMapping viewControllerHandlerMapping(...) {
    // 특정 URL 경로를 View 이름에 직접 매핑하는 HandlerMapping 생성
    // ...
}

@Bean
public HandlerMapping resourceHandlerMapping(...) {
    // 정적 리소스를 처리하는 HandlerMapping 생성
    // ...
}

...
```

즉 스프링 프레임워크 내부에서도, **새로운 종류의 Handler가 추가되면 코드를 수정해야 하는 구조**였습니다...! 신기하네요ㅎㅎ

### Best Practice

새롭게 알게 된 내용을 바탕으로 기존 제가 구현한 코드를 리팩터링 한다면 Factory 클래스를 하나 작성해서 Handler 관련 객체들의 생성 책임을 한 군데로 모으는 것도 좋을 것 같습니다. 다만 시간 관계상 여력이 된다면 다음 단계에서 같이 고민해 보겠습니다 🫡

yesjuhee replied:
관련해서 블로그 글 하나 작성했어요! 심심할 때 읽어보세요ㅎㅎ
https://yesjuhee.tistory.com/23

Byesol replied:
와우 정말 감탄했습니다. 😮  노랑의 정리한 글과 블로그 글 덕분에 정말 많이 배우고 갑니다. 

노랑이 보여준 factory 클래스를 보고 나니

스프링은 '새로운 핸들러가 추가될 때 내 코드 수정을 없앨 수 있을까?' 라는 질문에 답하려 한 게 아닌

'새로운 핸들러 추가와 같은' '변경'이 발생했을 때, 코드의 단 한 곳인 가장 예측 가능하고 안전한 단일 지점에서만 수정이 일어나도록 설계할 수 있을까?'

라는 질문에 답을 하기 위한 코드였다는 생각이 드네요.

또 스프링에서 Reflections를 이용한 완전 자동 스캔 방식을 사용하지 않았다면  어떤 트레이드오프가 있길래 스프링은 선택하지 않았을지도 궁금해지네요.


클래스패스를 모두 스캔하는 데 드는 성능 비용 때문?  어떤 빈이 등록될지 예측하기 어렵다?

이 주제에 대해서도 나중에 이야기 나눠보면 정말 재미있을 것 같습니다. 깊이 있는 분석과 공유 정말 고마워요! 👍

yesjuhee committed:
feat: NoHandlerFoundException 클래스 추가 및 예외 처리 개선

- 요청 처리 시 핸들러를 찾지 못하는 경우를 처리하기 위해 `NoHandlerFoundException` 클래스 추가
- `DispatcherServlet`에서 `NoHandlerFoundException` 발생 시 대응 로직 추가
- HTTP 404 에러 응답을 반환하는 `noHandlerFound` 메서드 추가
- 기존 코드에서 일반 `ServletException` 대신 `NoHandlerFoundException` 사용하도록 수정하여 예외 메시지 명확화

yesjuhee committed:
refactor: ControllerScanner 리플렉션 접근 제어 개선

- 기본 생성자 호출 시 `setAccessible(true)` 추가하여 리플렉션 호출 안정성 강화
- `getDeclaredConstructor` 메서드 분리로 코드 가독성 및 재사용성 향상
- 예외 메시지 및 로그를 통해 디버깅 용이성 확보
- 불필요한 개행 제거로 코드 정리

yesjuhee committed:
chore: AnnotationHandlerMapping 생성시 controller 패키지 위치 전달

yesjuhee commented:
벡터 리뷰 너무 좋은데요?! 생각해 볼 것이 많아서 많이 배웠습니다!
바로 머지 안해도 상관 없으니 추가적으로 이야기 나누고 싶은 부분이나, 궁금한 부분, 반박하고 싶은 부분 자유롭게 달아주셔도 됩니다 👍

Byesol reviewed (approved):
고생하셨어요~ 노랑! 
좋은 블로그 글 잘 읽었습니다! 덕분에 많이 배웠어요.

이만 merge 하도록 할게요. 다음 단계에서 만나요~~! ┌(・。・)┘♪
